src/transform/tests/testdata/canonicalize_mfp

```diff
@@ -24,3 +34,178 @@ build apply=CanonicalizeMfp
 | Get x (u0)
 | Filter false
 | Project (#1)
+
+# pushdown filters with a map
+build apply=CanonicalizeMfp
+(project
+    (filter
+        (map (get x) [(call_binary add_int64 #0 #1)])
+        [(call_binary lt #3 (1234 Int32))])
+     [3])
+----
+%0 =
+| Get x (u0)
+| Filter ((#0 + #1) < 1234)
+| Map (#0 + #1)
+| Project (#3)
+
+# pushdown filters without a map
+build apply=CanonicalizeMfp
+(project
+    (filter
+        (get x)
+        [(call_binary lt #0 (1234 Int32))])
+     [0 1])
+----
+%0 =
+| Get x (u0)
+| Filter (#0 < 1234)
+| Project (#0, #1)
+
+# retained filters with a map
+build apply=CanonicalizeMfp
+(project
+    (filter
+        (map (get x) [(call_variadic hmac_string [#0 #1 #2])])
+        [(call_unary is_null #3)])
+     [0 3])
+----
+%0 =
+| Get x (u0)
+| Map hmac(#0, #1, #2)
+| Filter isnull(#3)
+| Project (#0, #3)
+
+# multiple retained filters with a map. canonicalized filters should be re-sorted least to greatest
+build apply=CanonicalizeMfp
+(project
+    (filter
+        (map (get x) [(call_variadic hmac_string [#0 #1 #2]) (call_variadic hmac_string [#0 #1])])
+        [(call_unary is_null #4) (call_unary is_null #3)])
+     [0 3 4])
+----
+%0 =
+| Get x (u0)
+| Map hmac(#0, #1, #2), hmac(#0, #1)
+| Filter isnull(#3), isnull(#4)
+| Project (#0, #3, #4)
+
+# both pushdown and retained filters with a map
+build apply=CanonicalizeMfp
+(project
+    (filter
+        (map (get x) [(call_variadic hmac_string [#0 #1 #2])])
+        [
+            (call_binary lt #0 1234)
+            (call_unary is_null #3)
+        ])
+     [0 3])
+----
+%0 =
+| Get x (u0)
+| Filter (#0 < 1234)
+| Map hmac(#0, #1, #2)
+| Filter isnull(#3)
+| Project (#0, #3)
+
+# joining with duplicated pushdown predicates with maps. regression test for #10000
```

Comment:
I feel that the "joining" part is not relevant to what is happening in this test.

Here is what is happening in the example in #10000:

There's this sequence of projects and filters following an input of size 4 that you want to optimize:
```
mfp
4
project [0 1 0 3]
filter [(call_unary not (call_unary is_null #0)) 
    (call_binary or (call_binary eq #0 (5 Int32))
        (call_binary eq #0 (1337 Int32))) 
    (call_unary not (call_unary is_null #2)) 
    (call_binary or (call_binary eq #2 (5 Int32)) (call_binary eq #2 (1337 Int32)))
]
project [0 1 2 3]
project [0 1 3]
optimize
```

If you copy this into the [MFP unit tests](https://github.com/MaterializeInc/materialize/blob/main/src/expr/tests/testdata/mfp) and run the unit tests, you will see that you get:

```
map [!(isnull(#0)) ((#0 = 5) || (#0 = 1337))]
filter [#4 #4 #5 #5]
project [0 1 3]
```

The map is created via memoizing common subexpressions.

`isnull(#0)` can be inlined into a predicate, but `((#0 = 5) || (#0 = 1337))` cannot be,
so the `#4` predicates are `pushdown` but `#5` are `retained`.

```
filter [isnull(#0) isnull(#0)]
map [((#0 = 5) || (#0 = 1337))]
filter [#5 #5]
project [0 1 3]
```

The filters are deduplicated.

```
filter [isnull(#0)]
map [((#0 = 5) || (#0 = 1337))]
filter [#5]
project [0 1 3]
```

The map-filter-project sequence is re-optimized.
```
filter [isnull(#0)]
filter [((#0 = 5) || (#0 = 1337))]
project [0 1 3]
```

So I would describe this test as "regression test for #10000. Both pushdown and retained filters with a map. Even though there is no map in the test, the duplicated predicates cause the creation of a map via memoizing common subexpressions that then gets optimized away."

Now that I have explained what is actually happening in #10000, please add a test similar to what actually happens in the actual #10000 example where you can't tell that two predicates are duplicate just by looking at the filter and not the other parts of the map-filter-project sequence. In the #10000 example, the predicates become duplicate because the project maps column 0 into column 2, but you can also try out an example with a map, like the one below (assume the input only has two columns):
```
map (#0 + #1)
filter [((#0 + #1) = 5) (#2 = 5)] 
```